Feature: User can view annual leave

johnny-m-young opened the pull request:
## User story
```
As an employee of GDS
So that I can manage my remaining annual leave
I want to be able to see how many days of annual leave I have remaining this year.
```
## What
- Adds a `annual_leave_remaining` column to the `user` table.
- Implements the feature, with associated feature test

## Screenshots
### Before
![localhost_3000_(iPad Pro) (2)](https://github.com/jyoung-gds/govuk-holiday-logger/assets/94846816/2a6419ba-1652-4b3d-bd12-08c6a05f6ed8)
### After
![localhost_3000_(iPad Pro) (3)](https://github.com/jyoung-gds/govuk-holiday-logger/assets/94846816/f51fa389-7e9a-493d-8d55-dbcf78c977dd)
### Design specification
![localhost_3000_dashboard(iPad Pro) (2)](https://github.com/jyoung-gds/govuk-holiday-logger/assets/94846816/56d7388a-10de-4afa-a4bc-33de19e87526)

ryanb-gds commented on db/migrate/20230823160153_create_user_data.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateUserData < ActiveRecord::Migration[7.0]
+  def change
+    create_table :user_data do |t|
+      t.decimal :annual_leave_remaining, precision: 3, scale: 1, null: false
```
I think this is a good choice of data type for this column. Decimals are good for anything you might want to do maths with

ryanb-gds commented on app/models/user_data.rb:
```diff
@@ -0,0 +1,5 @@
+class UserData < ApplicationRecord
```
I would encourage you to try to think of a more specific name for this object if you can - maybe `AnnualLeaveYear`?

ollietreend replied:
+1 to what Ryan said.

Looking at the [entity relationship diagram](https://user-images.githubusercontent.com/94846816/262934836-076c4e4d-7f9d-439b-9fb3-fedadfbf6295.jpg) that you shared, which I realise covers more scope than just this PR, it did make me wonder:

What is the distinction between a `User` and a `UserData`? For example, why would the user's `staff_number` belong to one object, but their `years_of_service` belong to the other? The two seem very tightly coupled, which sort of indicates that they might not warrant being separate entities – it feels like they represent two halves of the same 'thing'.

It might be worth considering a couple of other options:

1. **Store all these fields on the `User` model.** So a user would have `annual_leave_remaining` directly on the `User` model.
2. **Define a clearer distinction between the two models.** For example, you might decide that the purpose of the `User` object is solely to facilitate users logging in to the app (e.g. it holds their login credentials, password reset tokens, and so on), and then have a `StaffMember` or `Employee` model to hold details specific to that user's employment (e.g. their staff number, years of service, line manager, remaining annual leave, etc.)

Option 2 might better support a future use-case where Staff Members could have multiple User accounts, or even to support different types of Users who can login but are not Staff Members. However if you [don't think you'll need that](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it), then Option 1 could be the simpler approach.

Happy to talk it through further if you want to! 🙂

johnny-m-young replied:
After extensive talks with Ollie, I've gone down the route of just having a `User` table and model and having everything contained in there (see new class diagram below). It is the assumption for the time being that all users will be employees, if this changes down the line (e.g. some users are not employees who will use the tool) then the annual leave data can be extracted into another class and migrated to another table, so this is a fairly low-risk assumption.

Thanks both for the input on this PR, it feels like quite a pivotal point in the project as decisions about data structure need to be made and it's important to get it right from the outset.

![class diagram (3)](https://github.com/jyoung-gds/govuk-holiday-logger/assets/94846816/9b4a7bb3-d4fd-4039-85dd-f25c86776cb2)

ryanb-gds reviewed (approved):
Looks good. Interested to see where you go with the rest of the dashboard build out

Commit:
Create DB migration for new attribute

Adds a database migration to add the
annual_leave_remaining column to the
user table.

A default value of 25.0 was used for this
new column because that is the amount of
annual leave a new starter has.

Commit:
Update the user factory

Adds the new annual_leave_remaining
attribute to the User factory

Commit:
Add a test for the feature

Adds a new feature test to check whether
a signed in user can view their remaining
annual leave on their dashboard.

Changes to the spec_helper.rb file are
needed in order to use the Devise helper
methods such as `sign_in` in RSpec feature
tests. They remove the deprecated
Devise::TestHelpers terminology and replace
it with up-to-date methods.

Commit:
Add annual leave remaining to root#index

This adds a panel to the root#index view
with the user's remaining annual leave displayed.
Feature test is passing and view is in line with
the design specification.
